### ISSUE TITLE
Add support for webhooks/callbacks

### ISSUE THREAD

**Issue by skjdghsdjgsdj** [babybuddy/babybuddy#813]
When we use Baby Buddy, we interact with it a couple ways: the web interface and the API through various integrations (Android app, custom devices, etc.). What would be nice is support for webhooks in the API.

Use case: we log a feeding one of three ways:
* Directly in the web interface
* Via the Android app
* Using a custom-built ESP32 device that uses the REST API

But, _regardless of how we log the feeding_, we'd like to get a webhook callback to an arbitrary HTTP(S) URL. That would let us set up a timer (outside of Baby Buddy) that notifies us after a certain time interval like two hours "hey, it's been a while, a feeding time is coming up shortly."

That's a specific use case around feeding, but a generic publish/subscribe mechanism would be nice for things like feedings, changes, etc. Another random example: a webhook around changes could let us separately keep track of likely number of diapers used, and then we get notified "hey, you've logged _n_ changes which probably means _n_ diapers and you might be running low."

I searched the docs and existing issues and didn't see this implemented or requested yet.

**Comment by cdubz** [comment]
I’m surprised there isn’t a more mature Django package for this. I’m open to this but it will take some research and consideration.

**Comment by skjdghsdjgsdj** [comment]
A potential idea, and shoot it down as you feel: MQTT support?

For example, for a simplistic initial use case, define an MQTT broker in a config file. Then if _any_ of the REST endpoints are hit, publish the endpoint and its payload to a user-defined (config file or something) MQTT broker. I have a Mosquitto instance available, and if Baby Buddy was just dumb--in a good way--and republishes the REST requests it gets to the MQTT broker, then another service can subscribe to the broker and take over from there. It would be loosely coupled and not require Baby Buddy to actually send its own outbound HTTP messages to arbitrary endpoints, and MQTT is by its nature a very light protocol.

Of course there could be information exposure, so after a basic implementation, maybe the MQTT relaying could be locked down to specific endpoints, users, etc., but it would be a nice starting point.

The de facto standard library for MQTT in Python seems to be https://pypi.org/project/paho-mqtt/ which I've used before and was pretty straightforward, at least as a subscriber

**Comment by cdubz** [comment]
Webhooks I have at least a passing familiarity with but MQTT is entirely new to me. This looks and sounds interesting but I'd probably have to lean on someone else contributing actual code here. As with webhooks -- I'm open to it, just can't say it's something I'll have the time to jump on. Though MQTT has piqued my curiosity for sure 😄

**Comment by cdubz** [comment]
Appreciate the support! And happy to help review code and move the idea forward. Will be a good way to learn about MQTT myself, hah.

**Comment by skjdghsdjgsdj** [comment]
I made an initial proof-of-concept of this. Keep in mind I'm a hobbyist with Python and this is the first time I've touched Django, so I'm probably violating all kinds of best design principles and this isn't ready for an actual PR. But I welcome your comments.

tl;dr: I made `DiaperChange` extend a new intermediate `BabyBuddyBaseModel` class which is abstract and itself extends `models.Model`. I overrode `save()` to call the parent to persist the data to the database first so IDs get populated, and then if there's an MQTT configuration in `production.py` (example config added to `production.example.py`), then Baby Buddy will publish the model's data as a JSON payload.

https://github.com/skjdghsdjgsdj/babybuddy/commit/cf46815c7d366273b62b93b46221bae7bd3c0849

Screenshot showing it ending up in MQTT:

![Screenshot 2024-11-01 at 10 33 57 AM](https://github.com/user-attachments/assets/7ed5f884-ff41-407b-823c-06d0ac4d418f)

I'm using [Mosquitto](https://mosquitto.org/) as an MQTT broker (server).

**Comment by cdubz** [comment]
@skjdghsdjgsdj this looks like a pretty interesting and straightforward start! Will you open a PR for the branch? You can set it as draft while we work through it.

For testing, it would also be helpful if you can give me some very basic instructions on how to set up the broker for testing.

Thanks for working on this!

**Comment by skjdghsdjgsdj** [comment]
I'll figure out the PR bit shortly. In the meantime, the high level approach you need to do is as follows. MQTT is a publish/subscribe mechanism. Clients publish messages to the MQTT "broker" (just a server) with a given topic and payload. Other clients can subscribe to that topic to listen for messages that got published. In this example, Baby Buddy publishes a message to the `babybuddy` topic with a JSON payload that contains the Django model's data.

* [Install an MQTT broker like Mosquitto](https://mosquitto.org/download/), like [this tutorial shows for Linux](https://docs.litmus.io/litmusedge/how-to-guides/integration-guides/install-mosquitto-mqtt-broker-ubuntu). You don't need to do the TLS steps in that tutorial. On macOS, [Mosquitto is available through Homebrew](https://formulae.brew.sh/formula/mosquitto) and I imagine the basic management commands are the same but with different config file paths.
* Add a block to `production.py` that looks like [the one I added in my copy of `production.example.py`](https://github.com/skjdghsdjgsdj/babybuddy/commit/cf46815c7d366273b62b93b46221bae7bd3c0849#diff-a6e06c452e2aaaa983fd35ec055ff51d08c0a62a6d32be2bd3b388b00b18b28b). You'll need to change the following, but `port`, `topic`, and `client_id` should all be okay just for testing:
  * `hostname` to something other than `localhost` if you're running Mosquitto on a different server than Baby Buddy
  * `username` and `password`, obviously, to what you defined when you set up Mosquitto
* Assuming you're using Linux or macOS, open a separate terminal running `mosquitto_sub` and listen to the topic `babybuddy`. On macOS, you can also use [MQTTX](https://mqttx.app/) which provides a more friendly UI for interacting with an MQTT broker.
* Launch Baby Buddy and try to save a diaper change. Hopefully, once the change gets saved, you should immediately see an MQTT message come through in the window running `mosquitto_sub` or in MQTTX.

My error handling is pretty bad so if it doesn't work, it'll be interesting at least.

My development environment involves running Baby Buddy in PyCharm with its built-in server. My MQTT broker is running on a separate Debian Linux server. Its `/etc/mosquitto/mosquitto.conf` looks like this:
```
pid_file /run/mosquitto/mosquitto.pid

persistence true
persistence_location /var/lib/mosquitto/

log_dest file /var/log/mosquitto/mosquitto.log

include_dir /etc/mosquitto/conf.d
password_file /etc/mosquitto/pwfile

listener 1883 192.168.1.193
```
This adds an optional dependency to [`paho-mqtt`](https://pypi.org/project/paho-mqtt/). I don't know how to set that up in `requirements.txt` or otherwise.

For future reference, [AWS IoT Core offers an MQTT broker as well](https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html) for people who don't want to self-host one.

**Comment by skjdghsdjgsdj** [comment]
PR: https://github.com/babybuddy/babybuddy/pull/894

**Comment by skjdghsdjgsdj** [comment]
As noted in #926: if this implementation included retained messages for most recent state (feedings, changes, etc.), then clients interested in real-time events via MQTT could receive the retained messages too instead of needing to first contact the REST API to get current state.

**Comment by skjdghsdjgsdj** [comment]
As a somewhat related followup: I was able to achieve this...kinda. You can have Baby Buddy's database replicate to a separate one, and then a third party script follows the replication log and translates events to JSON, which you can then publish.

You need the following:
* Baby Buddy configured to use MySQL as its database, not SQLite
* Apache Kafka, which you can [install from a Docker Compose stack](https://github.com/conduktor/kafka-stack-docker-compose)
* [Maxwell's daemon](https://github.com/zendesk/maxwell)

Once you have Kafka running, configure Maxwell's daemon to publish messages to Kafka. I filtered it to _only_ publish messages that happen in the Baby Buddy database instead of all databases. Then, you can get JSON messages of events that happen in the Baby Buddy database immediately instead of polling that look like this, as an example of an `UPDATE` query that ran:

```jsonc
{
  "database": "babybuddy",
  "table": "core_timer",
  "type": "update",
  "ts": 1753302801,
  "xid": 116807,
  "commit": true,
  "data": {
    "id": 369,
    "name": "sleep",
    "start": "2025-07-23 20:00:00.000000",
    "active": 1,
    "user_id": 7,
    "child_id": 1
  },
  "old": {
    "start": "2025-07-23 20:33:12.033780"
  }
}
```

Therefore, you can consume messages from Kafka in real time as things happen in the Baby Buddy database. It's obviously not perfect, like it can break if the schema changes, can be noisy, and pokes directly into the Baby Buddy internals, but it's better than polling. Kafka is also way, way more complex than MQTT given it's designed to scale for enterprises, not the typical home users like Baby Buddy.
